src/sinks/util/service/net/tcp.rs

```diff
@@ -0,0 +1,272 @@
+use std::{
+    net::SocketAddr,
+    task::{ready, Context, Poll},
+    time::Duration,
+};
+
+use futures::{future::BoxFuture, FutureExt};
+use snafu::{ResultExt, Snafu};
+use tokio::{
+    io::AsyncWriteExt,
+    net::{TcpSocket, TcpStream},
+    sync::oneshot,
+    time::sleep,
+};
+use tower::Service;
+
+use vector_config::configurable_component;
+use vector_core::tcp::TcpKeepaliveConfig;
+
+use crate::{
+    dns,
+    internal_events::{TcpSocketConnectionEstablished, TcpSocketOutgoingConnectionError},
+    net,
+    sinks::{util::retries::ExponentialBackoff, Healthcheck},
+};
+
+use super::HostAndPort;
+
+#[derive(Debug, Snafu)]
+pub enum TcpError {
+    #[snafu(display("Address was invalid: {}", reason))]
+    InvalidAddress { reason: String },
+
+    #[snafu(display("Failed to configure TCP socket: {}.", source))]
+    FailedToConfigure { source: std::io::Error },
+
+    #[snafu(display("Failed to send TCP message: {}", source))]
+    FailedToSend { source: std::io::Error },
+
+    #[snafu(display("Failed to connect to TCP endpoint: {}", source))]
+    FailedToConnect { source: std::io::Error },
+
+    #[snafu(display("No addresses returned."))]
+    NoAddresses,
+
+    #[snafu(display("Failed to resolve address: {}", source))]
+    FailedToResolve { source: crate::dns::DnsError },
+
+    #[snafu(display("Failed to get TCP stream back after send as channel closed unexpectedly."))]
+    ServiceStreamChannelClosed,
+}
+
+/// `TcpConnector` configuration.
+#[configurable_component]
+#[derive(Clone, Debug)]
+pub struct TcpConnectorConfig {
+    /// The address to connect to.
+    ///
+    /// Both IP addresses and hostnames/fully-qualified domain names are accepted formats.
+    ///
+    /// The address _must_ include a port.
+    address: HostAndPort,
```

Comment:
Nice. We should be using this everywhere it is applicable, which is a lot of components already ✅ auto doc generation converted.